mlflow/pyfunc/__init__.py

```diff
@@ -320,62 +433,192 @@ def predict(*args):
+                        defined in the ``__main__`` scope, the defining module should also be
+                        included in one of the listed locations.
+    :param artifacts: A dictionary containing ``<name, artifact_uri>`` entries. Remote artifact URIs
+                      will be resolved to absolute filesystem paths, producing a dictionary of
+                      ``<name, absolute_path>`` entries. ``python_model`` can reference these
+                      resolved entries as the ``artifacts`` property of the ``context`` attribute.
+                      For example, consider the following ``artifacts`` dictionary::
+
+                        {
+                            "my_file": "s3://my-bucket/path/to/my/file"
+                        }
+
+                      In this case, the ``"my_file"`` artifact will be downloaded from S3. The
+                      ``python_model`` can then refer to ``"my_file"`` as an absolute filesystem
+                      path via ``self.context.artifacts["my_file"]``.
+
+                      If *None*, no artifacts will be added to the model.
     """
-    if os.path.exists(dst_path):
-        raise Exception("Path '{}' already exists".format(dst_path))
-    os.makedirs(dst_path)
-    code = None
-    data = None
-    env = None
-
-    if data_path:
-        model_file = _copy_file_or_tree(src=data_path, dst=dst_path, dst_dir="data")
-        data = model_file
-
-    if code_path:
-        for path in code_path:
-            _copy_file_or_tree(src=path, dst=dst_path, dst_dir="code")
-        code = "code"
-
-    if conda_env:
-        shutil.copy(src=conda_env, dst=os.path.join(dst_path, "mlflow_env.yml"))
-        env = "mlflow_env.yml"
+    first_argument_set = {
+        "loader_module": loader_module,
+        "data_path": data_path,
+    }
+    second_argument_set = {
+        "artifacts": artifacts,
+        "python_model": python_model,
+    }
+    first_argument_set_specified = any([item is not None for item in first_argument_set.values()])
+    second_argument_set_specified = any([item is not None for item in second_argument_set.values()])
+    if first_argument_set_specified and second_argument_set_specified:
+        raise MlflowException(
+            message=(
+                "The following sets of parameters cannot be specified together: {first_set_keys}"
+                " and {second_set_keys}. All parameters in one set must be `None`. Instead, found"
+                " the following values: {first_set_entries} and {second_set_entries}".format(
+                    first_set_keys=first_argument_set.keys(),
+                    second_set_keys=second_argument_set.keys(),
+                    first_set_entries=first_argument_set,
+                    second_set_entries=second_argument_set)),
+            error_code=INVALID_PARAMETER_VALUE)
+    elif not (first_argument_set_specified or second_argument_set_specified):
```

Comment:
Is this logically sound? I seems like I could specify only `artifacts` and not get any error.